metricbeat/module/system/filesystem/helper.go

```diff
@@ -31,19 +36,56 @@ func GetFileSystemList() ([]sigar.FileSystem, error) {
 		return nil, err
 	}
 
-	// Ignore relative mount points, which are present for example
-	// in /proc/mounts on Linux with network namespaces.
-	filtered := fss.List[:0]
-	for _, fs := range fss.List {
-		if filepath.IsAbs(fs.DirName) {
+	return filterFileSystemList(fss.List), nil
+}
+
+// filterFileSystemList filters mountpoints to avoid virtual filesystems
+// and duplications
+func filterFileSystemList(fsList []sigar.FileSystem) []sigar.FileSystem {
+	var filtered []sigar.FileSystem
+	devices := make(map[string]sigar.FileSystem)
+	for _, fs := range fsList {
+		// Ignore relative mount points, which are present for example
+		// in /proc/mounts on Linux with network namespaces.
+		if !filepath.IsAbs(fs.DirName) {
+			debugf("Filtering filesystem with relative mountpoint %+v", fs)
+			continue
+		}
+
+		// Don't do further checks in special devices
+		if !filepath.IsAbs(fs.DevName) {
 			filtered = append(filtered, fs)
 			continue
 		}
-		debugf("Filtering filesystem with relative mountpoint %+v", fs)
+
+		// If the device name is a directory, this is a bind mount or nullfs,
+		// don't count it as it'd be counting again its parent filesystem.
+		devFileInfo, err := os.Stat(fs.DevName)
```

Comment:
One of the original reasons we added `ignore_types` was to avoid doing a `statfs` on autofs filesystems. Metricbeat was causing autofs mounts to never be unmounted because its continuous metric gathering was making autofs think the mount was being accessed.

Will this cause the same problem?

Reply:
Good point.
Here the stat is done against the device name, on #4823 the mentioned issued seemed to appear when doing statfs on the mount point (`DirName`). So I don't think it can cause the same problem.